src/main/scala/org/clulab/habitus/utils/ContextExtractor.scala

```diff
@@ -0,0 +1,260 @@
+      case None =>
+        //the combination of (sentid,freq) becomes the key for the value
+        sentidContext(key) = value
+    }
+  }
+  case class MostFreqEntity(sentId: Int, mention: String, mostFreqEntity: Option[String])
+
+
+  def createSentidContext(sentidContext: mutable.Map[Int,ContextDetails],
+                          mostFreqLocation0Sent:Seq[MostFreqEntity],
+                          mostFreqLocation1Sent:Seq[MostFreqEntity],
+                          mostFreqLocationOverall:Seq[MostFreqEntity],
+                          mostFreqDate0Sent:Seq[MostFreqEntity],
+                          mostFreqDate1Sent:Seq[MostFreqEntity],
+                          mostFreqDateOverall:Seq[MostFreqEntity],
+                          mostFreqCrop0Sent:Seq[MostFreqEntity],
+                          mostFreqCrop1Sent:Seq[MostFreqEntity],
+                          mostFreqCropOverall:Seq[MostFreqEntity],
+                          mostFreqFertilizer0Sent:Seq[MostFreqEntity],
+                          mostFreqFertilizer1Sent:Seq[MostFreqEntity],
+                          mostFreqFertilizerOverall:Seq[MostFreqEntity]
+                         ): Unit= {
+
+    //todo assert lengths of all the mostFreq* are same
+
+    //for each event mention, get the sentence id, and map it to a case class called contextDetails, which will have all of mostFreq* information
+    //note: zipping through only the list of one mostFreq* since all of them should have same lenghts.
+    for ((mostFreq, i) <- mostFreqLocation0Sent.zipWithIndex) {
+      checkSentIdContextDetails(sentidContext,mostFreq.sentId,
+        ContextDetails(mostFreq.mention,
+          checkIfEmpty(mostFreq.mostFreqEntity),
+          checkIfEmpty(mostFreqLocation1Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqLocationOverall(i).mostFreqEntity),
+          checkIfEmpty(mostFreqDate0Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqDate1Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqDateOverall(i).mostFreqEntity),
+          checkIfEmpty(mostFreqCrop0Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqCrop1Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqCropOverall(i).mostFreqEntity),
+          checkIfEmpty(mostFreqFertilizer0Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqFertilizer1Sent(i).mostFreqEntity),
+          checkIfEmpty(mostFreqFertilizerOverall(i).mostFreqEntity)
+        )
+      )
+    }
+  }
+  //if none, return "N/A", else return String value of entity e.g.:"SENEGAL"
+  def checkIfEmpty(mostFreqEntity: Option[String]): String = mostFreqEntity.getOrElse("N/A")
+
+  def findClosestNextLocation(mention: Mention, locations: Seq[Mention]): Mention = {
+    if (locations.length == 1) return locations.head
+    val nextLocations = locations.filter(_.tokenInterval.start > mention.arguments("value").head.tokenInterval.start)
+    if (nextLocations.nonEmpty) nextLocations.minBy(_.tokenInterval)
+    else null
```

Comment:
So here I'd very much recommend returning Option[Mention] as standard practice, for the benefit of others.